### PR TITLE
[IMP]: Stock - 1 usability improvement

### PR DESCRIPTION
**Pull request by atp-odoo**
[IMP]stock : Add three function field : reordering count and reording max and min qty count.
if there is only 1 reordering rule, display "min = Qty, max = Qty" with the correct quantity  instead of "Reordering Rule".

Task: https://www.odoo.com/web?#id=19408&view_type=form&model=project.task&menu_id=3942&action=327
Pad: https://pad.odoo.com/p/r.KqaU3MY4MnCnzIPV

### CODE REVIEW

File: addons/stock/product.py [hunk 1]
```diff
@@ -201,7 +201,22 @@ def _product_available_text(self, cr, uid, ids, field_names=None, arg=False, con
...
+                res[product.id].update({
+                    'reordering_min' : product.orderpoint_ids.product_min_qty,
+                    'reordering_max' : product.orderpoint_ids.product_max_qty
+                })
+        return res
+
     _columns = {
+        'reordering_count': fields.function(_reorder_rule_count, type='integer', multi=True),
+        'reordering_min': fields.function(_reorder_rule_count, type='float', multi=True),
+        'reordering_max': fields.function(_reorder_rule_count, type='float', multi=True),
```
Comment (atp-odoo) [inline addons/stock/product.py]: i think better to  give field name like 'reordering_max_qty' and 'reordering_min_qty'

File: addons/stock/product.py [hunk 2]
```diff
@@ -203,9 +203,15 @@ def _product_available_text(self, cr, uid, ids, field_names=None, arg=False, con
...
+        product_data = self.pool['stock.warehouse.orderpoint'].read_group(cr, uid, [('product_id', 'in', ids)], ['product_id', 'product_min_qty', 'product_max_qty'], ['product_id'], context=context)
         for data in product_data:
-            res[data['product_id'][0]] = int(data['product_id_count'])
+            res[data['product_id'][0]] = {'nbr_reordering_rules': int(data['product_id_count'])}
+            if int(data['product_id_count']) == 1:
+                res[data['product_id'][0]].update({
+                    'reordering_min_qty' : data['product_min_qty'],
+                    'reordering_max_qty' : data['product_max_qty']
+                })
+
```
Comment (dmo-odoo) [inline addons/stock/product.py]: As this is a computed field, I think you have to return the reordering_min_qty and reordering_max_qty even if product_id_count > 1. These field values make sense even if product_id_count > 1, it's just that you are going to use product_id_count to discriminate between one display or another in the view, but here ? You don't care, just return the fields informations.

File: addons/stock/product.py [hunk 3]
```diff
@@ -374,11 +382,19 @@ def _product_available_text(self, cr, uid, ids, field_names=None, arg=False, con
...
+            product_tmpl_id = data['__domain'][1][2][0]
+            res[product_tmpl_id]['nbr_reordering_rules'] += data['product_id_count']
+
         for data in product_data:
             product_tmpl_id = data['__domain'][1][2][0]
-            res[product_tmpl_id] += data['product_id_count']
+            if res[product_tmpl_id]['nbr_reordering_rules'] == 1:
+                res[product_tmpl_id]['reordering_min_qty'] += data['product_min_qty']
+                res[product_tmpl_id]['reordering_max_qty'] += data['product_max_qty']
+
```
Comment (dmo-odoo) [inline addons/stock/product.py]: Same as above. And additionally, why two loops ?

File: addons/stock/product_view.xml [hunk 4]
```diff
@@ -227,10 +227,21 @@
                         <field name="qty_available2" widget="statinfo"/>
                     </button>
                    <button class="oe_stat_button" name= "%(act_product_stock_move_open)d" type="action" attrs="{'invisible':[('type', '=', 'service')]}" string="Moves" groups="stock.group_stock_user" icon="fa-arrows-v"/>
-                   <button class="oe_stat_button" name="%(product_open_orderpoint)d" type="action" 
-                        attrs="{'invisible':[('type', '=', 'service')]}" icon="fa-refresh">
+                    <button class="oe_stat_button" name="%(product_open_orderpoint)d" type="action"
```
Comment (dmo-odoo) [inline addons/stock/product_view.xml]: Why is there an additional space at the start of the line now ?

File: addons/stock/product_view.xml [hunk 5]
```diff
@@ -248,12 +259,23 @@
                            type="object"  attrs="{'invisible':[('type', '=', 'service')]}" groups="stock.group_locations">
                         <field name="qty_available2" widget="statinfo"/>
                    </button>
-                   <button class="oe_stat_button" name= "action_view_stock_moves" type="object" attrs="{'invisible':[('type', '=', 'service')]}" string="Moves" groups="stock.group_stock_user" icon="fa-arrows-v"/>
-                   <button class="oe_stat_button" name="action_view_orderpoints" type="object" 
-                        attrs="{'invisible':[('type', 'in', ('service','consu'))]}" icon="fa-refresh">
+                    <button class="oe_stat_button" name= "action_view_stock_moves" type="object" attrs="{'invisible':[('type', '=', 'service')]}" string="Moves" groups= "stock.group_stock_user" icon="fa-arrows-v"/>
```
Comment (dmo-odoo) [inline addons/stock/product_view.xml]: Why is there an additional space at the start of the line and after the "groups=" now ?

File: addons/stock/product_view.xml [hunk 6]
```diff
@@ -248,12 +259,23 @@
...
+                    <button class="oe_stat_button" name="action_view_orderpoints" type="object"
```
Comment (dmo-odoo) [inline addons/stock/product_view.xml]: Why is there an additional space at the start of the line now ?

File: addons/stock/product.py (hunk 3 above)
Comment (atp-odoo) [inline addons/stock/product.py]: regarding two loops: i think this is due to master code. some of code for reording rules count already merged in master. i think this branch required rebase.

File: addons/stock/product.py [hunk 7]
```diff
@@ -202,10 +202,15 @@ def _product_available_text(self, cr, uid, ids, field_names=None, arg=False, con
...
+        res = dict.fromkeys(ids, {'nbr_reordering_rules': 0, 'reordering_min_qty': 0, 'reordering_max_qty': 0})
+        product_data = self.pool['stock.warehouse.orderpoint'].read_group(cr, uid, [('product_id', 'in', ids)], ['product_id', 'product_min_qty', 'product_max_qty'], ['product_id'], context=context)
         for data in product_data:
-            res[data['product_id'][0]] = int(data['product_id_count'])
+            res[data['product_id'][0]]['nbr_reordering_rules'] = int(data['product_id_count'])
+            res[data['product_id'][0]]['reordering_min_qty'] = data['product_min_qty']
+            res[data['product_id'][0]]['reordering_max_qty'] = data['product_max_qty']
+            if int(data['product_id_count']) == 1:
+                res[data['product_id'][0]].update({'reordering_min_qty': data['product_min_qty'], 
+                                                    'reordering_max_qty': data['product_max_qty']})
```
Comment (dmo-odoo) [inline addons/stock/product.py]: I think all the if part can be removed as you already assign those value above.

File: addons/stock/product.py [hunk 8]
```diff
@@ -374,11 +381,16 @@ def _product_available_text(self, cr, uid, ids, field_names=None, arg=False, con
...
-        product_data = self.pool['stock.warehouse.orderpoint'].read_group(cr, uid, [('product_id.product_tmpl_id', 'in', ids)], ['product_id'], ['product_id'], context=context)
+        res = dict.fromkeys(ids, {'nbr_reordering_rules': 0, 'reordering_min_qty': 0, 'reordering_max_qty': 0})
+        product_data = self.pool['stock.warehouse.orderpoint'].read_group(cr, uid, [('product_id.product_tmpl_id', 'in', ids)], ['product_id', 'product_min_qty', 'product_max_qty'], ['product_id'], context=context)
         for data in product_data:
             product_tmpl_id = data['__domain'][1][2][0]
-            res[product_tmpl_id] += data['product_id_count']
+            res[product_tmpl_id]['nbr_reordering_rules'] = int(data['product_id_count'])
+            res[product_tmpl_id]['reordering_min_qty'] = data['product_min_qty']
+            res[product_tmpl_id]['reordering_max_qty'] = data['product_max_qty']
+            if res[product_tmpl_id]['nbr_reordering_rules'] == 1:
```
Comment (dmo-odoo) [inline addons/stock/product.py]: I think all the if part can be removed as you already assign those value above.

Comment (atp-odoo) [inline addons/stock/product.py]: agree. not required if part. this was already hadled by view xml side using attrs. fixed

**Comment by dmo-odoo** [comment]
Merged at 3eed48ffcf796baa72e7eab68758434ec496be45.

**Pull request closed without merging**